README.md
# hsqoi

Haskell implementatin of the [QOI image format](https://phoboslab.org/log/2021/11/qoi-fast-lossless-image-compression).

## Benchmarks

The baseline is the [reference C code](https://github.com/phoboslab/qoi.git) @ [`e9069e`](https://github.com/phoboslab/qoi/commit/e9069e11a43d779b418679c7a50b2ec14f652085).

Taking a single big `.png` file (5616×3744) and taking a minimum of 5 runs gives, on Core i7 3930k @ 4.0 GHz:

| Implementation                   | Decoding, ms | , % of best C | Encoding, ms | , % of best C |
|----------------------------------|--------------|---------------|--------------|---------------|
| C, gcc 11, `-O3 -march=native`   | 228          | 117%          | 264          | 102%          |
| C, gcc 11, `-O3`                 | 230          | 118%          | 260          | 100%          |
| C, clang 13, `-O3 -march=native` | 211          | 108%          | 356          | 137%          |
| C, clang 13, `-O3`               | 195          | 100%          | 341          | 131%          |
| Haskell                          | 184          |  94%          | 240          |  90%          |

On Ryzen 3700X:

| Implementation                   | Decoding, ms | , % of best C | Encoding, ms | , % of best C |
|----------------------------------|--------------|---------------|--------------|---------------|
| C, gcc 11, `-O3 -march=native`   | 179          | 121%          | 198          | 101%          |
| C, gcc 11, `-O3`                 | 174          | 118%          | 196          | 100%          |
| C, clang 13, `-O3 -march=native` | 158          | 107%          | 266          | 136%          |
| C, clang 13, `-O3`               | 148          | 100%          | 252          | 129%          |
| Haskell                          | 148          | 100%          | 190          |  97%          |


ChangeLog.md
# Changelog for hsqoi

## Unreleased changes


test/Spec.hs
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE OverloadedLists #-}
{-# LANGUAGE Strict #-}

import qualified Data.Array.IArray as A
import qualified Data.Array.Unboxed as A
import qualified Data.ByteString as BS
import qualified Data.Vector as V
import Control.Monad
import Data.Binary.Combinators
import Data.Either
import Data.Word
import Test.Hspec
import Test.QuickCheck
import Test.Hspec.QuickCheck (modifyMaxSuccess)

import Data.Image.Qoi.Decoder
import Data.Image.Qoi.Encoder
import Data.Image.Qoi.Format
import Data.Image.Qoi.Pixel

newtype ShowAsBytes = ShowAsBytes { bytes :: BS.ByteString } deriving (Eq)

instance Show ShowAsBytes where
  show = show . toArray3 . bytes

data Image3 = Image3
  { iWidth :: Int
  , iHeight :: Int
  , iBytes :: ShowAsBytes
  } deriving (Eq, Show)

genPixel3 :: Gen Pixel3
genPixel3 = Pixel3 <$> chooseAny <*> chooseAny <*> chooseAny

genDiff3Bounded :: Word8 -> Pixel3 -> Gen Pixel3
genDiff3Bounded delta (Pixel3 r g b) = do
  diffs <- replicateM 3 $ choose (negate delta, delta - 1)
  case diffs of
       [dr, dg, db] -> pure $ Pixel3 (r + dr) (g + dg) (b + db)
       _ -> error "muh dependent types"

instance Arbitrary Image3 where
  arbitrary = do
    maxDim <- getSize
    width <- chooseInt (1, maxDim)
    height <- chooseInt (1, maxDim)

    px0 <- genPixel3

    pixels <- V.iterateNM (width * height) step px0
    pure $ Image3 width height $ ShowAsBytes $ BS.pack $ V.toList $ V.concatMap (\(Pixel3 r g b) -> [r, g, b]) pixels
    where
      step prevPixel = oneof [ pure prevPixel
                             , genDiff3Bounded 2 prevPixel
                             , genDiff3Bounded 8 prevPixel
                             , genDiff3Bounded 16 prevPixel
                             , genPixel3
                             ]

toArray3 :: BS.ByteString -> A.UArray Int Pixel3
toArray3 bs = A.array (0, pxCnt - 1) [ (i, Pixel3 r g b)
                                     | i <- [0..pxCnt - 1]
                                     , let r = bs `BS.index` (i * 3)
                                           g = bs `BS.index` (i * 3 + 1)
                                           b = bs `BS.index` (i * 3 + 2)
                                     ]
  where
    pxCnt = BS.length bs `div` 3

imgProperty :: Image3 -> IO ()
imgProperty Image3 { .. } = do
  let header = Header { hMagic = matchBytes
                      , hWidth = fromIntegral iWidth
                      , hHeight = fromIntegral iHeight
                      , hChannels = 3
                      , hColorspace = 0
                      }
  let encoded = encodeRaw header (bytes iBytes) 0
      decoded = decodeQoi $ BS.pack $ A.elems encoded
  decoded `shouldSatisfy` isRight
  case decoded of
       Left _ -> pure ()
       Right (header', Pixels3 pixels') -> do
         header' `shouldBe` header
         pixels' `shouldBe` toArray3 (bytes iBytes)
       Right _ -> fail "Expected Pixels3"

main :: IO ()
main = hspec $ modifyMaxSuccess (const 100000) $
  describe "QOI encoder" $
    parallel $ forM_ ([1..10] :: [Int]) $ \wrk ->
      it ("decode . encode = id (worker " <> show wrk <> ")") $ property imgProperty


app/Main.hs
{-# LANGUAGE Strict #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE RecordWildCards #-}

module Main where

import qualified Data.Array.Base as A
import qualified Data.Array.IO as A
import Codec.Picture.Png
import Codec.Picture.Types
import Data.Binary.Combinators
import Data.Functor
import GHC.Word
import System.Environment
import System.IO
import System.IO.Posix.MMap

import Data.Image.Qoi.Encoder
import Data.Image.Qoi.Format
import Data.Image.Qoi.Decoder
import Data.Image.Qoi.Pixel

toImage :: Header -> SomePixels -> DynamicImage
toImage Header { .. } (Pixels3 pixels) = ImageRGB8 $ generateImage f (fromIntegral hWidth) (fromIntegral hHeight)
  where
    f x y = let Pixel3 r g b = pixels A.! (y * fromIntegral hWidth + x)
             in PixelRGB8 r g b
toImage Header { .. } (Pixels4 pixels) = ImageRGBA8 $ generateImage f (fromIntegral hWidth) (fromIntegral hHeight)
  where
    f x y = let Pixel4 r g b a = pixels A.! (y * fromIntegral hWidth + x)
             in PixelRGBA8 r g b a

dumpUArray :: FilePath -> A.UArray Int Word8 -> IO ()
dumpUArray file arr = withFile file WriteMode $ \h -> do
  ioarr <- A.unsafeThaw arr
  A.hPutArray h ioarr (A.numElements arr)

main :: IO ()
main = getArgs >>=
  \case ["decode", inFile] -> do
           bs <- unsafeMMapFile inFile
           case decodeQoi bs of
                Right (header, pixels) -> do print header
                                             case pixels of
                                                  Pixels3 pixels' -> print $ A.bounds pixels'
                                                  Pixels4 pixels' -> print $ A.bounds pixels'
                Left err -> putStrLn $ "Unable to decode: " <> show err
        ["decode", inFile, outFile] -> do
           bs <- unsafeMMapFile inFile
           case decodeQoi bs of
                Left err -> putStrLn $ "Unable to decode: " <> show err
                Right (header, pixels) -> void $ writeDynamicPng outFile $ toImage header pixels
        ["encode_raw", inFile, width, height] -> do
           bs <- unsafeMMapFile inFile
           print $ A.bounds $ encodeRaw (Header matchBytes (read width) (read height) 3 0) bs 0
        ["encode_raw", inFile, width, height, outFile] -> do
           bs <- unsafeMMapFile inFile
           dumpUArray outFile $ encodeRaw (Header matchBytes (read width) (read height) 3 0) bs 0
        _ -> putStrLn "Wrong usage"


src/Data/Image/Qoi/Util.hs
{-# LANGUAGE Strict #-}
{-# LANGUAGE BinaryLiterals #-}

module Data.Image.Qoi.Util where

import qualified Data.Array.Base as A
import qualified Data.ByteString as BS
import qualified Data.ByteString.Internal as BSI
import Data.Bits
import Data.Word
import Foreign.ForeignPtr
import Foreign.Storable (peekByteOff)

infix 8 .>>., .<<.
(.>>.), (.<<.) :: Bits a => a -> Int -> a
(.>>.) = shiftR
(.<<.) = shiftL
{-# INLINE (.<<.) #-}
{-# INLINE (.>>.) #-}

infix 4 !
(!) :: BS.ByteString -> Int -> Word8
(BSI.PS x _ _) ! i = BSI.accursedUnutterablePerformIO $ withForeignPtr x $ \p -> peekByteOff p i
{-# INLINE (!) #-}

unsafeShrink :: A.STUArray s Int Word8 -> Int -> A.STUArray s Int Word8
unsafeShrink arr@(A.STUArray l _ n marr) cnt
  | cnt >= n = arr
  | otherwise = A.STUArray l (l + cnt - 1) cnt marr


src/Data/Image/Qoi/Pixel.hs
{-# LANGUAGE Strict #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE MagicHash #-}
{-# LANGUAGE UnboxedTuples #-}
{-# LANGUAGE NondecreasingIndentation #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE QuantifiedConstraints #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE BangPatterns #-}
{-# OPTIONS_GHC -O2 -fllvm #-}

module Data.Image.Qoi.Pixel where

import qualified Data.Array.Base as A
import Control.Monad.ST
import Data.Word
import GHC.Base
import GHC.ST
import GHC.Word

data Pixel3 = Pixel3 Word8 Word8 Word8 deriving (Show, Eq)
data Pixel4 = Pixel4 Word8 Word8 Word8 Word8 deriving (Show, Eq)

instance A.MArray (A.STUArray s) Pixel3 (ST s) where
  getBounds (A.STUArray l u _ _) = pure (l, u)
  {-# INLINE getBounds #-}
  getNumElements (A.STUArray _ _ n _) = pure n
  {-# INLINE getNumElements #-}

  newArray_ arrBounds = A.newArray arrBounds (Pixel3 0 0 0)
  {-# INLINE newArray_ #-}
  unsafeNewArray_ (l, u) = A.unsafeNewArraySTUArray_ (l, u) (*# 3#)
  {-# INLINE unsafeNewArray_ #-}

  unsafeRead (A.STUArray _ _ _ marr#) (I# n#) = ST $ \s1# ->
    let n'# = n# *# 3#
        !(# s2#, r# #) = readWord8Array# marr# n'#         s1#
        !(# s3#, g# #) = readWord8Array# marr# (n'# +# 1#) s2#
        !(# s4#, b# #) = readWord8Array# marr# (n'# +# 2#) s3#
     in (# s4#, Pixel3 (W8# r#) (W8# g#) (W8# b#) #)
  {-# INLINE unsafeRead #-}
  unsafeWrite (A.STUArray _ _ _ marr#) (I# n#) (Pixel3 (W8# r#) (W8# g#) (W8# b#)) = ST $ \s1# ->
    let n'# = n# *# 3#
        s2# = writeWord8Array# marr# n'#         r# s1#
        s3# = writeWord8Array# marr# (n'# +# 1#) g# s2#
        s4# = writeWord8Array# marr# (n'# +# 2#) b# s3#
     in (# s4#, () #)
  {-# INLINE unsafeWrite #-}

instance A.IArray A.UArray Pixel3 where
  bounds (A.UArray l u _ _) = (l, u)
  {-# INLINE bounds #-}
  numElements (A.UArray  _ _ n _) = n
  {-# INLINE numElements #-}
  unsafeArray lu ies = runST (A.unsafeArrayUArray lu ies $ Pixel3 0 0 0)
  {-# INLINE unsafeArray #-}
  unsafeAt (A.UArray _ _ _ arr#) (I# n#) = Pixel3 (W8# (indexWord8Array# arr# n'#))
                                                  (W8# (indexWord8Array# arr# (n'# +# 1#)))
                                                  (W8# (indexWord8Array# arr# (n'# +# 2#)))
    where
      n'# = n# *# 3#
  {-# INLINE unsafeAt #-}

instance A.MArray (A.STUArray s) Pixel4 (ST s) where
  getBounds (A.STUArray l u _ _) = pure (l, u)
  {-# INLINE getBounds #-}
  getNumElements (A.STUArray _ _ n _) = pure n
  {-# INLINE getNumElements #-}

  newArray_ arrBounds = A.newArray arrBounds (Pixel4 0 0 0 0)
  {-# INLINE newArray_ #-}
  unsafeNewArray_ (l, u) = A.unsafeNewArraySTUArray_ (l, u) (*# 4#)
  {-# INLINE unsafeNewArray_ #-}

  unsafeRead (A.STUArray _ _ _ marr#) (I# n#) = ST $ \s1# ->
    let n'# = n# *# 4#
        !(# s2#, r# #) = readWord8Array# marr# n'#         s1#
        !(# s3#, g# #) = readWord8Array# marr# (n'# +# 1#) s2#
        !(# s4#, b# #) = readWord8Array# marr# (n'# +# 2#) s3#
        !(# s5#, a# #) = readWord8Array# marr# (n'# +# 3#) s4#
     in (# s5#, Pixel4 (W8# r#) (W8# g#) (W8# b#) (W8# a#) #)
  {-# INLINE unsafeRead #-}
  unsafeWrite (A.STUArray _ _ _ marr#) (I# n#) (Pixel4 (W8# r#) (W8# g#) (W8# b#) (W8# a#)) = ST $ \s1# ->
    let n'# = n# *# 4#
        s2# = writeWord8Array# marr# n'#         r# s1#
        s3# = writeWord8Array# marr# (n'# +# 1#) g# s2#
        s4# = writeWord8Array# marr# (n'# +# 2#) b# s3#
        s5# = writeWord8Array# marr# (n'# +# 3#) a# s4#
     in (# s5#, () #)
  {-# INLINE unsafeWrite #-}

instance A.IArray A.UArray Pixel4 where
  bounds (A.UArray l u _ _) = (l, u)
  {-# INLINE bounds #-}
  numElements (A.UArray  _ _ n _) = n
  {-# INLINE numElements #-}
  unsafeArray lu ies = runST (A.unsafeArrayUArray lu ies $ Pixel4 0 0 0 0)
  {-# INLINE unsafeArray #-}
  unsafeAt (A.UArray _ _ _ arr#) (I# n#) = Pixel4 (W8# (indexWord8Array# arr# n'#))
                                                  (W8# (indexWord8Array# arr# (n'# +# 1#)))
                                                  (W8# (indexWord8Array# arr# (n'# +# 2#)))
                                                  (W8# (indexWord8Array# arr# (n'# +# 3#)))
    where
      n'# = n# *# 4#
  {-# INLINE unsafeAt #-}

class (Eq a, forall s. A.MArray (A.STUArray s) a (ST s)) => Pixel a where
  addRGB  :: a -> Word8 -> Word8 -> Word8 -> a
  addRGBA :: a -> Word8 -> Word8 -> Word8 -> Word8 -> a

  toRGBA :: a -> (Word8, Word8, Word8, Word8)
  fromRGBA :: Word8 -> Word8 -> Word8 -> Word8 -> a

instance Pixel Pixel3 where
  addRGB (Pixel3 r g b) dr dg db = Pixel3 (r + dr) (g + dg) (b + db)
  {-# INLINE addRGB #-}
  addRGBA (Pixel3 r g b) dr dg db _ = Pixel3 (r + dr) (g + dg) (b + db)
  {-# INLINE addRGBA #-}
  toRGBA (Pixel3 r g b) = (r, g, b, 255)
  {-# INLINE toRGBA #-}
  fromRGBA r g b _ = Pixel3 r g b
  {-# INLINE fromRGBA #-}

instance Pixel Pixel4 where
  addRGB (Pixel4 r g b a) dr dg db = Pixel4 (r + dr) (g + dg) (b + db) a
  {-# INLINE addRGB #-}
  addRGBA (Pixel4 r g b a) dr dg db da = Pixel4 (r + dr) (g + dg) (b + db) (a + da)
  {-# INLINE addRGBA #-}
  toRGBA (Pixel4 r g b a) = (r, g, b, a)
  {-# INLINE toRGBA #-}
  fromRGBA r g b a = Pixel4 r g b a
  {-# INLINE fromRGBA #-}


src/Data/Image/Qoi/Decoder.hs
{-# LANGUAGE Strict #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE BinaryLiterals #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# OPTIONS_GHC -O2 -fllvm #-}

module Data.Image.Qoi.Decoder
( decodeQoi
, DecodeError(..)
, SomePixels(..)
) where

import qualified Data.Array.Base as A
import qualified Data.Array.ST as A
import qualified Data.ByteString as BS
import qualified Data.ByteString.Lazy as BSL
import Control.Monad
import Data.Binary
import Data.Binary.Get
import Data.Bits

import Data.Image.Qoi.Format
import Data.Image.Qoi.Pixel
import Data.Image.Qoi.Util

data ChunkResult pixel
  = One pixel
  | Repeat pixel Int
  | Lookback Int
  | Stop

peekChunk :: Pixel pixel => BS.ByteString -> Int -> pixel -> (Int, ChunkResult pixel)
peekChunk str pos prevPixel
  | byte .>>. 6 == 0     = (1, Lookback $ fromIntegral $ byte .&. 0b00111111)
  | byte .>>. 5 == 0b010 = (1, Repeat prevPixel $ fromIntegral $ 1 + byte .&. 0b00011111)
  | byte .>>. 5 == 0b011 = (2, Repeat prevPixel $ 33 + (fromIntegral (byte .&. 0b00011111) .<<. 8
                                                    .|. fromIntegral (str ! pos + 1)
                                                       )
                           )
  | byte .>>. 6 == 0b10  = let dr = (byte .>>. 4 .&. 0b11) - 2
                               dg = (byte .>>. 2 .&. 0b11) - 2
                               db = (byte        .&. 0b11) - 2
                            in (1, One $ addRGB prevPixel dr dg db)
  | byte .>>. 5 == 0b110 = let next = str ! pos + 1
                               dr = (byte .&. 0b00011111) - 16
                               dg = (next .>>. 4)         - 8
                               db = (next .&. 0b00001111) - 8
                            in (2, One $ addRGB prevPixel dr dg db)
  | byte .>>. 4 == 0b1110 = let threeBytes :: Word32
                                threeBytes = fromIntegral byte .<<. 16
                                         .|. fromIntegral (str ! pos + 1) .<<. 8
                                         .|. fromIntegral (str ! pos + 2)
                                dr = fromIntegral (threeBytes .>>. 15 .&. 0b11111) - 16
                                dg = fromIntegral (threeBytes .>>. 10 .&. 0b11111) - 16
                                db = fromIntegral (threeBytes .>>. 5  .&. 0b11111) - 16
                                da = fromIntegral (threeBytes         .&. 0b11111) - 16
                             in (3, One $ addRGBA prevPixel dr dg db da)
  | byte .>>. 4 == 0b1111 = let hr = byte .>>. 3 .&. 0b1
                                hg = byte .>>. 2 .&. 0b1
                                hb = byte .>>. 1 .&. 0b1
                                ha = byte        .&. 0b1
                                (r, g, b, a) = toRGBA prevPixel
                                r' = (negate hr .&. (str ! pos + 1))
                                 .|. (hr - 1)   .&. r
                                g' = (negate hg .&. (str ! pos + 1 + fromIntegral hr))
                                 .|. (hg - 1)   .&. g
                                b' = (negate hb .&. (str ! pos + 1 + fromIntegral (hr + hg)))
                                 .|. (hb - 1)   .&. b
                                a' = (negate ha .&. (str ! pos + 1 + fromIntegral (hr + hg + hb)))
                                 .|. (ha - 1)   .&. a
                             in (1 + fromIntegral (hr + hg + hb + ha), One $ fromRGBA r' g' b' a')
  | otherwise = (0, Stop)
  where
    byte = str ! pos

decodePixels :: Pixel pixel => BS.ByteString -> Int -> Int -> A.UArray Int pixel
decodePixels str strFrom n = A.runSTUArray $ do
  (mvec :: A.STUArray s Int pixel) <- A.unsafeNewArray_ (0, n - 1)

  running <- A.newArray @(A.STUArray s) (0, 63 :: Int) (fromRGBA 0 0 0 255)

  let step inPos outPos prevPixel
        | outPos < n = do
            let (diff, chunk) = peekChunk str inPos prevPixel
            case chunk of
                 One px        -> do A.unsafeWrite mvec outPos px
                                     updateRunning running px
                                     step (inPos + diff) (outPos + 1)   px
                 Lookback pos  -> do px <- A.unsafeRead running pos
                                     A.unsafeWrite mvec outPos px
                                     step (inPos + diff) (outPos + 1)   px
                 Repeat px cnt -> do forM_ [0..cnt - 1] $ \i -> A.unsafeWrite mvec (outPos + i) px
                                     updateRunning running px
                                     step (inPos + diff) (outPos + cnt) px
                 Stop          -> pure outPos
        | otherwise = pure outPos
  finish <- step strFrom 0 (fromRGBA 0 0 0 255)

  forM_ [finish .. n - 1] $ \i -> A.unsafeWrite mvec i (fromRGBA 0 0 0 255)

  pure mvec

data SomePixels where
  Pixels3 :: A.UArray Int Pixel3 -> SomePixels
  Pixels4 :: A.UArray Int Pixel4 -> SomePixels
  deriving (Show)

data DecodeError
  = HeaderError String
  | UnsupportedChannels Int
  | UnpaddedFile
  deriving (Show)

decodeWHeader :: BS.ByteString
              -> Either (BSL.ByteString, ByteOffset, String) (BSL.ByteString, ByteOffset, Header)
              -> Either DecodeError (Header, SomePixels)
decodeWHeader _ (Left (_, _, err)) = Left $ HeaderError err
decodeWHeader str (Right (_, consumed, header))
  | any (\i -> (str ! BS.length str - i) /= 0) [1..4] = Left UnpaddedFile
  | hChannels header == 3 = Right (header, Pixels3 decode')
  | hChannels header == 4 = Right (header, Pixels4 decode')
  | otherwise = Left $ UnsupportedChannels $ fromIntegral $ hChannels header
  where
    decode' :: Pixel pixel => A.UArray Int pixel
    decode' = decodePixels str (fromIntegral consumed) (fromIntegral $ hWidth header * hHeight header)

decodeQoi :: BS.ByteString -> Either DecodeError (Header, SomePixels)
decodeQoi str = decodeWHeader str $ decodeOrFail $ BSL.fromStrict $ BS.take 14 str


src/Data/Image/Qoi/Encoder.hs
{-# LANGUAGE Strict #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE BinaryLiterals #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE FlexibleContexts #-}
{-# OPTIONS_GHC -O2 -fllvm #-}

module Data.Image.Qoi.Encoder
( encodeRaw
) where

import Prelude hiding (Maybe(..))

import qualified Data.Array.Base as A
import qualified Data.Array.ST as A
import qualified Data.ByteString as BS
import qualified Data.ByteString.Lazy as BSL
import Control.Applicative
import Control.Monad
import Control.Monad.ST
import Data.Binary
import Data.Bits
import Data.Functor
import Data.Maybe

import Data.Image.Qoi.Pixel
import Data.Image.Qoi.Util
import Data.Image.Qoi.Format
import Data.Proxy

type VarEncoder s = A.STUArray s Int Word8 -> Int -> Maybe (ST s Int)

-- `isBounded n b` checks that -b <= n < b
isBounded :: Word8 -> Word8 -> Bool
isBounded d h = d + h < 2 * h
{-# INLINE isBounded #-}

encodeDiff8 :: Word8 -> Word8 -> Word8 -> Word8 -> VarEncoder s
encodeDiff8 dr dg db 0 out outPos
  | (`isBounded` 2) `all` [dr, dg, db] = let byte = 0b10000000 .|. ((dr + 2) .<<. 4)
                                                               .|. ((dg + 2) .<<. 2)
                                                               .|.  (db + 2)
                                          in Just $ A.unsafeWrite out outPos byte $> 1
encodeDiff8 _ _ _ _ _ _  = Nothing
{-# INLINE encodeDiff8 #-}

encodeDiff16 :: Word8 -> Word8 -> Word8 -> Word8 -> VarEncoder s
encodeDiff16 dr dg db 0 out outPos
  | uncurry isBounded `all` [(dr, 16), (dg, 8), (db, 8)] = let b1 = 0b11000000 .|. (dr + 16)
                                                               b2 = (dg + 8) .<<. 4
                                                                .|. (db + 8)
                                                            in Just $ do A.unsafeWrite out outPos       b1
                                                                         A.unsafeWrite out (outPos + 1) b2
                                                                         pure 2
encodeDiff16 _ _ _ _ _ _  = Nothing
{-# INLINE encodeDiff16 #-}

encodeDiff24 :: Word8 -> Word8 -> Word8 -> Word8 -> VarEncoder s
encodeDiff24 dr dg db da out outPos
  | (`isBounded` 16) `all` [dr, dg, db, da] = let bytes :: Word32
                                                  bytes = (0b11100000 .<<. 16)
                                                      .|. fromIntegral (dr + 16) .<<. 15
                                                      .|. fromIntegral (dg + 16) .<<. 10
                                                      .|. fromIntegral (db + 16) .<<. 5
                                                      .|. fromIntegral (da + 16)
                                               in Just $ do A.unsafeWrite out outPos       (fromIntegral $ bytes .>>. 16)
                                                            A.unsafeWrite out (outPos + 1) (fromIntegral $ bytes .>>. 8)
                                                            A.unsafeWrite out (outPos + 2) (fromIntegral   bytes)
                                                            pure 3
  | otherwise = Nothing
{-# INLINE encodeDiff24 #-}

encodeIndex :: Eq pixel => pixel -> Word8 -> pixel -> VarEncoder s
encodeIndex px hash runningPx out outPos
  | px == runningPx = Just $ A.unsafeWrite out outPos hash $> 1
  | otherwise = Nothing
{-# INLINE encodeIndex #-}

encodeColor :: Pixel pixel => pixel -> pixel -> VarEncoder s
encodeColor px1 px0 out outPos = Just $ do
  A.unsafeWrite out outPos bh
  when (hr == 1) $ A.unsafeWrite out (outPos + 1) r1
  when (hg == 1) $ A.unsafeWrite out (outPos + 1 + hr) g1
  when (hb == 1) $ A.unsafeWrite out (outPos + 1 + hr + hg) b1
  when (ha == 1) $ A.unsafeWrite out (outPos + 1 + hr + hg + hb) a1
  pure (1 + hr + hg + hb + ha)
  where
    (r1, g1, b1, a1) = toRGBA px1
    (r0, g0, b0, a0) = toRGBA px0
    hr = fromEnum $ r1 /= r0
    hg = fromEnum $ g1 /= g0
    hb = fromEnum $ b1 /= b0
    ha = fromEnum $ a1 /= a0
    bh = 0b11110000 .|. (fromIntegral hr .<<. 3)
                    .|. (fromIntegral hg .<<. 2)
                    .|. (fromIntegral hb .<<. 1)
                    .|.  fromIntegral ha
{-# INLINE encodeColor #-}

maxRunLen :: Int
maxRunLen = 8224

encodeRun :: Int -> A.STUArray s Int Word8 -> Int -> ST s Int
encodeRun runLen out outPos
  | runLen == 0 = pure outPos
  | runLen <= 32 = A.unsafeWrite out outPos (0b01000000 .|. fromIntegral (runLen - 1)) $> outPos + 1
  | otherwise = do let runLen' = runLen - 33
                   A.unsafeWrite out outPos       (0b01100000 .|. fromIntegral (runLen' .>>. 8))
                   A.unsafeWrite out (outPos + 1) (fromIntegral runLen')
                   pure $ outPos + 2
{-# INLINE encodeRun #-}

maxResultSize :: Header -> (Int, BS.ByteString)
maxResultSize h@Header { .. } = (maxLen, headerBS)
  where
    headerBS = BSL.toStrict $ encode h
    maxLen = BS.length headerBS
           + fromIntegral (hChannels + 1) * fromIntegral (hWidth * hHeight)
           + 4 -- end padding

encodeIntoArray :: forall pixel s. Pixel pixel
                => Proxy pixel
                -> Int
                -> BS.ByteString
                -> Int
                -> A.STUArray s Int Word8
                -> ST s Int
encodeIntoArray _ headerLen inBytes startPos result = do
  running <- A.newArray @(A.STUArray s) (0, 63 :: Int) (fromRGBA @pixel 0 0 0 255)

  let step inPos runLen prevPx outPos
        | inPos + 3 <= inLen
        , readPixel inBytes inPos == prevPx =
          if runLen /= maxRunLen - 1
             then step (inPos + 3) (runLen + 1) prevPx outPos
             else encodeRun maxRunLen result outPos >>= step (inPos + 3) 0 prevPx
        | inPos + 3 <= inLen = do
          let (r0, g0, b0, a0) = toRGBA prevPx
          let px = readPixel inBytes inPos
          let (r1, g1, b1, a1) = toRGBA px
          let (dr, dg, db, da) = (r1 - r0, g1 - g0, b1 - b0, a1 - a0)
          outPos' <- encodeRun runLen result outPos

          let hash = pixelHash px
          pxDiff <- case encodeDiff8 dr dg db da result outPos' of
                         Just act -> act
                         _ -> do runningPx <- A.unsafeRead running hash
                                 fromJust $ encodeIndex px (fromIntegral hash) runningPx result outPos'
                                        <|> encodeDiff16 dr dg db da result outPos'
                                        <|> encodeDiff24 dr dg db da result outPos'
                                        <|> encodeColor px prevPx result outPos'

          A.unsafeWrite running hash px

          step (inPos + 3) 0 px (outPos' + pxDiff)
        | runLen /= 0 = encodeRun runLen result outPos
        | otherwise = pure outPos

  step startPos 0 (fromRGBA 0 0 0 255) headerLen
  where
    inLen = BS.length inBytes
{-# INLINE encodeIntoArray #-}

encodeRaw :: Header -> BS.ByteString -> Int -> A.UArray Int Word8
encodeRaw header inBytes startPos = A.runSTUArray $ do
  result <- A.unsafeNewArray_ (0, maxLen - 1)
  forM_ [0 .. headerLen - 1] $ \i -> A.unsafeWrite result i (headerBS ! i)
  final <- if hChannels header == 3
              then encodeIntoArray @Pixel3 Proxy headerLen inBytes startPos result
              else encodeIntoArray @Pixel4 Proxy headerLen inBytes startPos result
  forM_ [0..3] $ \i -> A.unsafeWrite result (final + i) 0
  pure $ unsafeShrink result (final + 4)
  where
    (maxLen, headerBS) = maxResultSize header
    headerLen = BS.length headerBS
